src/DynamoPackages/PackageLoader.cs

```diff
@@ -418,16 +418,51 @@ public Package ScanPackageDirectory(string directory, bool checkCertificates)
                 // prevent loading unsigned packages if the certificates are required on package dlls
                 if (checkCertificates)
                 {
-                    CheckPackageNodeLibraryCertificates(directory, discoveredPkg);
+                    CheckPackageNodeLibraryCertificates(directory, discoveredPackage);
                 }
 
-                // prevent duplicates
-                if (LocalPackages.All(pkg => pkg.Name != discoveredPkg.Name))
+                var discoveredVersion = CheckAndGetPackageVersion(discoveredPackage.VersionName, discoveredPackage.Name, discoveredPackage.RootDirectory);
+
+                var existingPackage = LocalPackages.FirstOrDefault(package => package.Name == discoveredPackage.Name);
+
+                // Is this a new package?
+                if (existingPackage == null)
+                {
+                    // Yes
+                    Add(discoveredPackage);
+                    return discoveredPackage; // success
+                }
+
+                var existingVersion = CheckAndGetPackageVersion(existingPackage.VersionName, existingPackage.Name, existingPackage.RootDirectory);
+
+                // Is this a duplicated package?
+                if (existingVersion == discoveredVersion)
                 {
-                    Add(discoveredPkg);
-                    return discoveredPkg; // success
+                    // Duplicated with the same version
+                    throw new LibraryLoadFailedException(directory, String.Format(Properties.Resources.DulicatedPackage, discoveredPackage.Name, discoveredPackage.RootDirectory));
+                } 
+                
+                // Is the existing version newer?
+                if (existingVersion > discoveredVersion)
+                {
+                    // Newer version already exist
+                    Log(String.Format(Properties.Resources.DuplicatedOlderPackage, discoveredPackage.Name, discoveredPackage.RootDirectory, existingVersion.ToString(), discoveredVersion.ToString()), WarningLevel.Moderate);
+                    return null;
+                }
+
+                // Is the existing package already loaded?
+                if (existingPackage.Loaded)
```

Comment:
What does this case mean?

Reply:
I will add more comments.

Reply:
Why do we have a separate code path for this case? Shouldn't we continue to line 475 even for the case where the user has added a new package path?